tungsten_ci_utils/nexus_cleanup/docker_cleanup.groovy

```diff
@@ -0,0 +1,118 @@
+    def retentionDate = DateTime.now().minusHours(repoToClean.retentionHours);
+    log.info(":::Cleanup script started for ${repositoryName} with the retention date ${retentionDate}");
+
+    MaintenanceService service = container.lookup("org.sonatype.nexus.repository.maintenance.MaintenanceService");
+    def repo = repository.repositoryManager.get(repositoryName);
+    def tx = repo.facet(StorageFacet.class).txSupplier().get();
+    try {
+        tx.begin();
+        def assets = tx.browseAssets(tx.findBucket(repo));
+        tx.rollback()
+        int deletedComponentCount = 0
+        int skippedComponentCount = 0
+        def keepingComponents = [:];
+        ImmutableList.copyOf(assets).each{ asset ->
+            if( asset.componentId() != null ) {
+                def ch = asset.attributes().child("content");
+                //log.info("ch=${ch}");
+                //log.info("a.m=${asset.getEntityMetadata()}");
+                //def ldo = asset.lastDownloaded();
+                def last_modified = toDateTime(ch.get("last_modified",Date.class));
+                if(  last_modified >= retentionDate )  {
+                    log.debug("Asset ${asset.name()} last_modified ${last_modified} is after ${retentionDate}, keeping ${asset.componentId()}");
+                    keepingComponents[asset.componentId()]=true;
+                } else if( keepingComponents[asset.componentId()] ) {
+                    log.info("Asset ${asset.name()} last_modified ${last_modified} is ignored as it belongs to ${asset.componentId()} which is kept");
+                } else {
+                    tx.begin();
+                    def comp = tx.findComponentInBucket(asset.componentId(),tx.findBucket(repo));
+                    tx.rollback();
+                    if( comp != null ) {
+                        if (comp.lastUpdated() < retentionDate) {
+                            log.info("Deleting ${repositoryName}/${comp.name()}:${comp.version()}, changed ${comp.lastUpdated()}, pushed ${last_modified}: both before ${retentionDate}");
+                            if(!dryRun) {
+                                service.deleteComponent(repo, comp);
+                            }
+                            deletedComponentCount++;
+                        } else {
+                            log.info("Skipping ${repositoryName}/${comp.name()}:${comp.version()}, changed ${comp.lastUpdated()}: pushed ${last_modified} after ${retentionDate}");
+                            skippedComponentCount++;
+                        }
+                    }
+                }
+            }
+        }
+        log.info("Deleted ${deletedComponentCount} docker image-tags from ${repositoryName}");
+        log.info("Keeping ${keepingComponents.size()} image-tags which had recent assets");
+        log.info("Skipped ${skippedComponentCount} recent image-tags which had old assets");
+
+        //ImmutableList.copyOf(components).each { comp ->
```

Comment:
delete if not needed

Reply:
It's not a mature code, there is a good chance this fragment will be needed - where do I keep this fragment?